src/libstd/num/f64.rs

```diff
@@ -546,7 +567,27 @@ impl f64 {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     pub fn log2(self) -> f64 {
-        unsafe { intrinsics::log2f64(self) }
+        if !cfg!(target_os = "sunos") {
+            unsafe { intrinsics::log2f64(self) }
+        } else {
+            // Illumos requires a wrapper around the log2 function because of
+            // its non-standard behavior
+            if self.is_finite() {
+                if self > 0.0 {
+                    unsafe { intrinsics::log2f64(self) }
+                } else if self == 0.0 {
+                    NEG_INFINITY // log2(0) = -Inf
+                } else {
+                    NAN // log2(-n) = NaN
+                }
+            } else if self.is_nan() {
+                self // log2(NaN) = NaN
+            } else if self > 0.0 {
+                self // log2(Inf) = Inf
+            } else {
+                NAN // log2(-Inf) = NaN
+            }
```

Comment:
Ah sorry I was just indicating that the macro could probably be removed in favor of generic functions and/or otherwise non-meta-programming pieces. Could the duplication between the functions here be reduced?